basestar-database-server/src/main/java/io/basestar/database/action/UpdateAction.java

```diff
@@ -80,76 +80,89 @@ public Instance after(final ValueContext valueContext, final Context expressionC
         final Map<String, Object> data = new HashMap<>();
         final UpdateOptions.Mode mode = Nullsafe.orDefault(options.getMode(), UpdateOptions.Mode.REPLACE);
 
-        final long version;
+        final long afterVersion;
+        final Long beforeVersion;
         final Instant created;
         final Instant updated = ISO8601.now();
 
         if (before == null) {
-            if(mode == UpdateOptions.Mode.CREATE) {
-                version = 1L;
+            if (mode == UpdateOptions.Mode.CREATE) {
+                afterVersion = 1L;
                 created = updated;
             } else {
                 throw new ObjectMissingException(options.getSchema(), id);
             }
         } else {
-            if(!schema.getQualifiedName().equals(Instance.getSchema(before))) {
+            if (!schema.getQualifiedName().equals(Instance.getSchema(before))) {
                 throw new IllegalStateException("Cannot change instance schema");
             }
 
-            if(mode == UpdateOptions.Mode.MERGE || mode == UpdateOptions.Mode.MERGE_DEEP) {
+            if (mode == UpdateOptions.Mode.MERGE || mode == UpdateOptions.Mode.MERGE_DEEP) {
                 data.putAll(before);
             }
 
-            final Long beforeVersion = Instance.getVersion(before);
+            beforeVersion = Instance.getVersion(before);
             assert beforeVersion != null;
 
-            if(options.getVersion() != null && !beforeVersion.equals(options.getVersion())) {
+            if (options.getVersion() != null && !beforeVersion.equals(options.getVersion())) {
                 throw new VersionMismatchException(options.getSchema(), id, options.getVersion());
             }
 
-            version = beforeVersion + 1;
+            afterVersion = beforeVersion + 1;
             created = Instance.getCreated(before);
         }
 
-        if(options.getData() != null) {
-            if(mode == UpdateOptions.Mode.MERGE_DEEP) {
+        if (options.getData() != null) {
+            if (mode == UpdateOptions.Mode.MERGE_DEEP) {
                 mergeDeep(data, options.getData());
             } else {
                 data.putAll(options.getData());
             }
         }
-        if(options.getExpressions() != null) {
+        if (options.getExpressions() != null) {
             options.getExpressions().forEach((k, expr) -> data.put(k, expr.evaluate(expressionContext)));
         }
         final Map<String, Object> initial = new HashMap<>(schema.create(valueContext, data, null));
 
         Instance.setId(initial, id);
         Instance.setSchema(initial, schema.getQualifiedName());
-        Instance.setVersion(initial, version);
+
         Instance.setCreated(initial, created);
-        Instance.setUpdated(initial, updated);
+        if (before != null && !updatedValuesPresent(before, initial)) {
```

Comment:
Can you use this existing method instead of using updatedValuesPresent:

`schema.areEqual(before, initial)`

Reasoning for this is that types can define their own equality rules, and this is the single place where equality of instances should be defined.